pep-9999.rst

```diff
@@ -0,0 +1,188 @@
+is supported by both f-strings and ``str.format()``, but not %-formatting.
+
+Synopsis:
+
+.. code-block:: pycon
+
+    >>> x = -.00001
+    >>> f'{x:z.1f}'
+    '0.0'
+
+    >>> x = decimal.Decimal('-.00001')
+    >>> '{}'.format('+z.1f')
+    '+0.0'
+
+    >>> x = 0
+    >>> f'{x:z.1f}'  # not supported on int type
+    ValueError: Negative zero coercion (z) not allowed in int format specifier
+
+    >>> x = -.00001
+    >>> '%z.1f' % x  # not supported on %-formatting
+    ValueError: unsupported format character 'z' (0x7a) at index 1
+
+.. _`Format Specification Mini-Language`: https://docs.python.org/3/library/string.html#format-specification-mini-language
+
+
+Rationale
+=========
+
+There are use cases where negative zero is unwanted in formatted number
+output -- arguably it's the more common case.  Expanding the format
+specification is the best way to support this because number formatting
+already incorporates rounding, and the normalization of negative zero must
+happen after rounding.
+
+While it is possible to pre-round and normalize a number before formatting,
+it's tedious and prone to error if the rounding doesn't precisely match
+that of the format spec.  Furthermore, functions that wrap formatting would
+find themselves having to parse format specs to extract the precision
+information.  For example, consider how this utility for formatting
+one-dimensional numerical arrays would be complicated by such pre-rounding:
+
+.. code-block:: python
+
+    def format_vector(v, format_spec='8.2f'):
+        """Format a vector (any iterable) using given per-term format string."""
+        return f"[{','.join(f'{term:{format_spec}}' for term in v)}]"
+
+The solution must be opt-in, because we can't change the behavior of
+programs that may be expecting or relying on negative zero when formatting
+numbers.
+
+The proposed extension is intentionally ``[sign][z]`` rather than
+``[sign[z]]``, since the latter would defeat ``sign``'s default (``-``)
+when using the ``z`` option.  Precisely because it is the default, ``-`` is
+not well known, and requiring it would increase the difficulty for those
+using or encountering the new option.
+
+There is a long-standing decision not to enhance the old %-formatting, so
+only f-strings and ``str.format()`` support the new option.
+
+
+Similar Work
+============
+
+The same formatting option for negative zero has been `proposed for
+C++ std::format()`_.  While the proposal was withdrawn for C++20, a
+consensus proposal is promised for C++23.  (For what it's worth, the original
+feature request prompting this PEP was argued without knowledge of the
+C++ proposal.)
+
+.. _`proposed for C++ std::format()`: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1496r2.pdf
+
+
+Reference Implementation
+========================
+
+A reference implementation exists at `pull request #30049`_.
+
+.. _`pull request #30049`: https://github.com/python/cpython/pull/30049
```

Comment:
Whilst this is in draft I think this is fine, but we recently changed several links to permalinks -- that might be considered at a later stage (@CAM-Gerlach may have a view here)

Reply:
I'm afraid I don't follow, sorry—that's a link to a pull request, which already is a permalink (that PR will always map to that number), not a link to a specific file as per the previous discussion.

Reply:
The contents of the branch that the PR represents might change, though? 

A

Reply:
Yes, but that's intentional, as the link is to the current reference implementation, not to an out of date snapshot of it (which would need to constantly be updated every time the user made a commit), just like if it were the link to a GitHub/GitLab repository that contained a reference implementation independent of CPython itself, as on many PEPs. Additionally, being a PR, even if the fork branch backing it is moved, deleted or modified further, once it is merged (or closed, etc) it will perpetually display the commits, files, etc. from that point, as it isn't actually showing the user's branch but rather the ref stored under `pull/$ID/head:$BRANCHNAME` of the upstream repo.

Reply:
Sounds good -- flagged as a potential issue but seems no such issue exists, which is always nice!
A

Reply:
Better safe than sorry, as we say across the pond!